src/Ide/Plugin/Eval.hs

```diff
@@ -0,0 +1,299 @@
+import           Development.IDE.Core.RuleTypes (GetModSummary (..),
+                                                 GhcSessionDeps (..))
+import           Development.IDE.Core.Shake     (use_)
+import           Development.IDE.GHC.Util       (evalGhcEnv, hscEnv,
+                                                 textToStringBuffer)
+import           Development.IDE.Types.Location (toNormalizedFilePath',
+                                                 uriToFilePath')
+import           DynamicLoading                 (initializePlugins)
+import           GHC
+import           GHC.Generics                   (Generic)
+import           GhcMonad                       (modifySession)
+import           GhcPlugins                     (defaultLogActionHPutStrDoc,
+                                                 gopt_set, gopt_unset,
+                                                 interpWays, updateWays,
+                                                 wayGeneralFlags,
+                                                 wayUnsetGeneralFlags)
+import           HscTypes
+import           Ide.Plugin
+import           Ide.Types
+import           Language.Haskell.LSP.Core      (LspFuncs (getVirtualFileFunc))
+import           Language.Haskell.LSP.Types
+import           Language.Haskell.LSP.VFS       (VirtualFile (..))
+import           PrelNames                      (pRELUDE)
+import           System.IO                      (IOMode (WriteMode), hClose, openFile)
+import           System.IO.Extra                (newTempFile)
+
+descriptor :: PluginId -> PluginDescriptor
+descriptor plId =
+  (defaultPluginDescriptor plId)
+    { pluginId = plId,
+      pluginCodeLensProvider = Just provider,
+      pluginCommands = [evalCommand]
+    }
+
+extractMatches :: Maybe Text -> [([(Text, Int)], Range)]
+extractMatches = goSearch 0 . maybe [] T.lines
+  where
+    checkMatch = T.stripPrefix "-- >>> "
+    looksLikeSplice l
+      | Just l' <- T.stripPrefix "--" l
+      = not (" >>>" `T.isPrefixOf` l')
+      | otherwise
+      = False
+
+    goSearch _ [] = []
+    goSearch line (l : ll)
+      | Just match <- checkMatch l =
+        goAcc (line + 1) [(match, line)] ll
+      | otherwise =
+        goSearch (line + 1) ll
+
+    goAcc line acc [] = [(reverse acc,Range p p)] where p = Position line 0
+    goAcc line acc (l:ll)
+      | Just match <- checkMatch l =
+        goAcc (line + 1) ([(match, line)] <> acc) ll
+      | otherwise =
+        (reverse acc,r) : goSearch (line + 1) ll
+      where
+        r = Range p p'
+        p = Position line 0
+        p' = Position (line + spliceLength) 0
+        spliceLength = length (takeWhile looksLikeSplice (l:ll))
+
+provider :: CodeLensProvider
+provider lsp _state plId CodeLensParams {_textDocument} = response $ do
+  let TextDocumentIdentifier uri = _textDocument
+  contents <- liftIO $ getVirtualFileFunc lsp $ toNormalizedUri uri
+  let text = Rope.toText . (_text :: VirtualFile -> Rope.Rope) <$> contents
```

Comment:
I think it might be better to use `virtualFileText` from `Language.Haskell.LSP.VFS` and delete extra dependency. It looks like the function is defined exactly in this way.

Reply:
Thanks. resolved